controller/backup_controller.go

```diff
@@ -116,6 +121,49 @@ func (bc *BackupController) enqueueBackupForMonitor(key string) {
 	bc.queue.Add(key)
 }
 
+func (bc *BackupController) enqueueInstanceManagerChange(obj interface{}) {
+	im, isInstanceManager := obj.(*longhorn.InstanceManager)
+	if !isInstanceManager {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("received unexpected obj: %#v", obj))
+			return
+		}
+
+		// use the last known state, to enqueue, dependent objects
+		im, ok = deletedState.Obj.(*longhorn.InstanceManager)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("DeletedFinalStateUnknown contained invalid object: %#v", deletedState.Obj))
+			return
+		}
+	}
+
+	imType, err := datastore.CheckInstanceManagerType(im)
+	if err != nil || imType != longhorn.InstanceManagerTypeEngine {
+		return
+	}
+
+	if im.Status.OwnerID != bc.controllerID {
+		return
+	}
+
+	if im.Status.CurrentState == longhorn.InstanceManagerStateStarting || im.Status.CurrentState == longhorn.InstanceManagerStateRunning {
+		return
+	}
```

Comment:
I am a bit confused about this condition. Assume that we have the following case:
* instance-manager-e-xxx is running on node-1
* instance-manager-e-xxx is failed on node-1 and remains to be failed
* instance-manager-e-xxx still has im.Status.OwnerID is node-1
* only node-1 handles this instance manager because of the check 
  ```
  if im.Status.OwnerID != bc.controllerID {
		  return
	  }
  ```
* we enqueue all the backup with backup.Status.OwnerID is node-1
* Latter on, we reject to handle all of those backups because this condition `currentNodeInstanceManagerAvailable := bc.ds.IsDefaultEngineInstanceManagerAvailability(bc.controllerID)`. So, is it true that the backups are still not being handled by anyone?

I think we don't observer this problem yet because of the 30s resync period that dumps all backup CRs into other nodes. So, those nodes have chance to take over the backups. However, if we disable resync, we might see the problem. How do you think?

Reply:
You are right, thanks for pointing it out. Fixed.